File: a349(command_interpreter)/C++/command_interpreter.cpp
#include <iostream>
#include <cstdio>

int main() {
  int memory[8] = {0},
      buffer[8] = {0};
  for (int i = 0; i < 8; i++) {
    std::cin >> memory[i];
  }
  int num_of_command;
  std::cin >> num_of_command;
  for (int i = 0; i < num_of_command; i++) {
    std::string com;
    std::cin >> com;
    switch (com[0]) {
      case'L': {
        int rd, ms;
        std::cin >> rd >> ms;
        buffer[rd] = memory[ms];
        break;
      }
      case 'S': {
        int md, rs;
        std::cin >> md >> rs;
        memory[md] = buffer[rs];
        break;
      }
      case 'A': {
        int rd, rs1, rs2;
        std::cin >> rd >> rs1 >> rs2;
        buffer[rd] = buffer[rs1] + buffer[rs2];
        break;
      }
      case 'M': {
        int rd, rs;
        std::cin >> rd >> rs;
        buffer[rd] = buffer[rs];
        break;
      }
    }
  }
  std::cout << buffer[0] << '\n'
            << memory[0] << '\n';

  return 0;
}

File: a349(command_interpreter)/Python/command_interpreter.py
from sys import stdin, stdout

memory = [0 for _ in range(8)]
buffer = [0 for _ in range(8)]

for i in range(8):
    init_val = int(stdin.readline())
    memory[i] = init_val
num_of_command = int(stdin.readline())
for _ in range(num_of_command):
	format = stdin.readline().split(' ')
	command = format[0]
	if command[0] == 'L':
		rd = int(format[1])
		ms = int(format[2])
		buffer[rd] = memory[ms]
	elif command[0] == 'S':
		md = int(format[1])
		rs = int(format[2])
		memory[md] = buffer[rs]
	elif command[0] == 'A':
		rd = int(format[1])
		rs1 = int(format[2])
		rs2 = int(format[3])
		buffer[rd] = buffer[rs1] + buffer[rs2]
	else:
		rd = int(format[1])
		rs = int(format[2])
		buffer[rd] = buffer[rs]
stdout.write(f'{buffer[0]}\n{memory[0]}\n')
